### PR TITLE
feat: [SNOW-3317601] fix column names on joins with duplicates

### PR DESCRIPTION
**Pull request by sfc-gh-jwilkowski**
Fixes: #2846 

### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.
   * [ ] I've described my changes in the documentation.

### Changes description
https://snowflakecomputing.atlassian.net/browse/SNOW-3317601

### CODE REVIEW

File: src/snowflake/cli/api/output/types.py [hunk 1]
```diff
@@ -138,6 +140,29 @@ def _prepare_payload(self, cursor: SnowflakeCursor | DictCursor):
             return (k for k in cursor)
```
Comment (sfc-gh-olorek) [inline src/snowflake/cli/api/output/types.py:145]: Heads up: this `DictCursor` branch yields row dicts straight from the connector, which has already collapsed duplicate keys before `QueryResult` ever sees them. So for `DictCursor` callers, `self.column_names` can now say `['id', 'id_2']` while the actual row dicts still have only a single `id` key — `column_names` and row payloads silently disagree.

In practice `DictCursor` is only used for metadata queries (`show objects`, `describe`, etc.) where collisions don't happen, so it's unlikely to bite. But worth either extending the fix to rebuild the dicts here using the uniquified names, or dropping a short comment calling out that the asymmetry is intentional.

File: src/snowflake/cli/api/output/types.py [hunk 2]
```diff
@@ -138,6 +140,29 @@ def _prepare_payload(self, cursor: SnowflakeCursor | DictCursor):
...
         return ({k: v for k, v in zip(self.column_names, row)} for row in cursor)
 
+    @staticmethod
+    def _uniquify_column_names(column_names: list[str]) -> list[str]:
+        """Disambiguate duplicate column names by appending a counter suffix."""
+        unique_names: list[str] = []
+        used_names: set[str] = set(column_names)
```
Comment (sfc-gh-olorek) [inline src/snowflake/cli/api/output/types.py:152]: Nit: consider a one-line comment here explaining why `used_names` is pre-seeded with the full input. This is the subtle bit that makes the `['id', 'id', 'id_2']` → `['id', 'id_3', 'id_2']` case work (the real `id_2` further down reserves that name so the duplicate `id` skips past it). The test at `test_query_result_uniquify_does_not_rename_real_column_matching_generated_suffix` locks the behavior in, but a reader looking at this function in isolation has to reconstruct the reasoning. Not blocking.

**Review by sfc-gh-olorek — COMMENTED** [review]
Nice focused fix. Two inline notes — the DictCursor one is worth thinking about before merge, the other is a nit.

File: src/snowflake/cli/api/output/types.py (hunk 1 above)
Comment (sfc-gh-jwilkowski) [reply src/snowflake/cli/api/output/types.py:145]: added

File: src/snowflake/cli/api/output/types.py (hunk 2 above)
Comment (sfc-gh-jwilkowski) [reply src/snowflake/cli/api/output/types.py:152]: I'd not go into explaining such details. Overall intent of this function is to make columns unique, and that's what the name and docstring hint at. Tests will catch potential issues

**Review by sfc-gh-olorek — DISMISSED** [review]
Thanks for the comment on DictCursor — addresses the asymmetry concern. Fine leaving the uniquify internals as-is; tests cover the edge case.

**Review by sfc-gh-olorek — APPROVED** [review]
